### PR TITLE
[BugFix] Solve column reuse problem

### PR DESCRIPTION
**Pull request by liuyehcf**
## Background

```sql
select *, row_number() over(partition by v1 order by cast(v3 as bigint)) from t0
```

the original type of `v3` is already bigint, so the cast(v3 as bigint)) will lead to a column reuse. 

## Main works of this pr:

1. Introduce `ColumnReuseGuarantor` to wrap reused `ColumnRefOperator` with `CloneOperator`
2. Introduce `ColumnReuseChecker` to check if there is reused column at plan stage.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 12 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|12|00.00%|[229, 230, 231, 232, 233, 234, 357, 358, 359, 360, 361, 362]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 12 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|12|00.00%|[229, 230, 231, 232, 233, 234, 357, 358, 359, 360, 361, 362]|

**Comment by trueeyu** [comment]
Add one test case?

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 12 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|12|00.00%|[229, 230, 231, 232, 233, 234, 357, 358, 359, 360, 361, 362]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 12 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|12|00.00%|[229, 230, 231, 232, 233, 234, 357, 358, 359, 360, 361, 362]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 12 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|12|00.00%|[229, 230, 231, 232, 233, 234, 357, 358, 359, 360, 361, 362]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 100 / 109 (91.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalProjectOperator.java|0|1|00.00%|[42]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/ProjectImplementationRule.java|0|1|00.00%|[40]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneProjectRule.java|0|1|00.00%|[61]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|2|5|40.00%|[428, 436, 489]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|3|5|60.00%|[53, 138]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|3|4|75.00%|[178]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateToExternalTableScanRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnExpressionToChildProject.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/UniquenessBasedTablePruneRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2AnalyticRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyAggProjectFilterRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctDataSkewEliminateRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/PlanValidator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CboTablePruneRule.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnEnforcer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneGroupByKeysRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ExistentialApply2OuterJoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CPJoinGardener.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateScanRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyAggFilterRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinOrder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/ReorderJoinRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalProjectOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteDuplicateAggregateFnRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SemiReorderRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneUnionEmptyRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneProjectColumnsRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 19 / 23 (82.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|4|50.00%|[485, 486]|
|:large_blue_circle:|com/starrocks/common/util/ColumnReuseUtil.java|17|19|89.47%|[27, 36]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 57 / 72 (79.17%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/ColumnReuseGuarantor.java|29|37|78.38%|[36, 46, 85, 86, 87, 92, 93, 94]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/ColumnReuseChecker.java|25|32|78.12%|[52, 81, 82, 83, 88, 89, 90]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/PlanValidator.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ColumnReuseGuarantor.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                if (valueColumnRefs.contains(valueColumnRef)) {
+                    reinsertKeyColumnRefs.add(entry.getKey());
+                } else {
+                    valueColumnRefs.add(valueColumnRef);
+                }
+            }
+
+            for (ColumnRefOperator key : reinsertKeyColumnRefs) {
+                ColumnRefOperator value = projectMap.remove(key).cast();
+                projectMap.put(key, new CloneOperator(value));
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ColumnReuseGuarantor.java:64]: directly put can replace the old value.

Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ColumnReuseGuarantor.java:64]: I need to get the old one(return from the remove method), and wrap it with CloneOperator.

**Review by packy92 — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 57 / 72 (79.17%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/ColumnReuseGuarantor.java|29|37|78.38%|[36, 46, 85, 86, 87, 92, 93, 94]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/ColumnReuseChecker.java|25|32|78.12%|[52, 81, 82, 83, 88, 89, 90]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/PlanValidator.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 30 / 37 (81.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/ColumnReuseChecker.java|25|32|78.12%|[52, 81, 82, 83, 88, 89, 90]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/PlanValidator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/CloneDuplicateColRefRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneUnionEmptyRule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/partition/partition_hash_map.h|0|2|00.00%|[231, 356]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ColumnReuseGuarantor.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ColumnReuseGuarantor.java]: Fixed

**Review by packy92 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**